Source/WebCore/workers/service/background-fetch/BackgroundFetch.cpp

```diff
@@ -0,0 +1,351 @@
+    , m_registrationKey(fetch.m_registrationKey)
+    , m_request(WTFMove(request))
+    , m_index(index)
+{
+}
+
+BackgroundFetch::Record::~Record()
+{
+    auto callbacks = std::exchange(m_responseCallbacks, { });
+    for (auto& callback : callbacks)
+        callback(makeUnexpected(ExceptionData { TypeError, "Record is gone"_s }));
+
+    auto bodyCallbacks = std::exchange(m_responseBodyCallbacks, { });
+    for (auto& callback : bodyCallbacks)
+        callback(makeUnexpected(ResourceError { errorDomainWebKitInternal, 0, { }, "Record is gone"_s }));
+}
+
+bool BackgroundFetch::Record::isMatching(const ResourceRequest& request, const CacheQueryOptions& options) const
+{
+    return DOMCacheEngine::queryCacheMatch(request, m_request.internalRequest, m_response, options);
+}
+
+BackgroundFetchRecordInformation BackgroundFetch::Record::information() const
+{
+    return BackgroundFetchRecordInformation { m_identifier, m_request.internalRequest, m_request.options, m_request.guard, m_request.httpHeaders, m_request.referrer };
+}
+
+void BackgroundFetch::Record::complete(const CreateLoaderCallback& createLoaderCallback)
+{
+    ASSERT(!m_loader);
+    // FIXME: Handle Range headers
+    m_loader = createLoaderCallback(*this, ResourceRequest { m_request.internalRequest }, FetchOptions { m_request.options }, m_fetch->m_origin);
+    if (!m_loader)
+        abort();
+}
+
+void BackgroundFetch::Record::abort()
+{
+    if (m_isAborted)
+        return;
+
+    m_isAborted = true;
+
+    auto callbacks = std::exchange(m_responseCallbacks, { });
+    for (auto& callback : callbacks)
+        callback(makeUnexpected(ExceptionData { AbortError, "Background fetch was aborted"_s }));
+
+    auto bodyCallbacks = std::exchange(m_responseBodyCallbacks, { });
+    for (auto& callback : bodyCallbacks)
+        callback(makeUnexpected(ResourceError { errorDomainWebKitInternal, 0, { }, "Background fetch was aborted"_s, ResourceError::Type::Cancellation }));
+
+    if (!m_loader)
+        return;
+    m_loader->abort();
+    m_loader = nullptr;
+}
+
+void BackgroundFetch::Record::didSendData(uint64_t size)
+{
+    if (m_fetch)
+        m_fetch->didSendData(size);
+}
+
+void BackgroundFetch::Record::didReceiveResponse(ResourceResponse&& response)
+{
+    m_response = response;
+    auto callbacks = std::exchange(m_responseCallbacks, { });
+    for (auto& callback : callbacks)
+        callback(ResourceResponse { m_response });
+    if (m_fetch)
+        m_fetch->storeResponse(m_index, WTFMove(response));
+}
+
+void BackgroundFetch::Record::didReceiveResponseBodyChunk(const SharedBuffer& data)
+{
+    m_responseDataSize += data.size();
+    if (m_fetch)
+        m_fetch->storeResponseBodyChunk(m_index, data);
+
+    if (!m_responseBodyCallbacks.isEmpty()) {
+        auto buffer = SharedBuffer::create(data.data(), data.size());
+        for (auto& callback : m_responseBodyCallbacks)
+            callback(RefPtr { buffer.ptr() });
```

Comment:
`callback(buffer.copyRef())` should work I think

Reply:
Will switch buffer to RefPtr